src/helper/bit-tools/shape-tool.js

```diff
@@ -0,0 +1,222 @@
+        if ((!this.shape || !this.shape.isInserted()) &&
+                selectedItems && selectedItems.length === 1 && selectedItems[0].shape === this.shapeType) {
+            // Infer that an undo occurred and get back the active shape
+            this.shape = selectedItems[0];
+            if (this.shape.data.zoomLevel !== paper.view.zoom) {
+                this.shape.strokeWidth = this.shape.strokeWidth / this.shape.data.zoomLevel * paper.view.zoom;
+                this.shape.data.zoomLevel = paper.view.zoom;
+                this.thickness = this.shape.strokeWidth;
+            }
+            this.filled = this.shape.strokeWidth === 0;
+            const color = this.filled ? this.shape.fillColor : this.shape.strokeColor;
+            this.color = color ? color.toCSS() : null;
+        } else if (this.shape && this.shape.isInserted() && !this.shape.selected) {
+            // Shape got deselected
+            this.commitShape();
+        }
+    }
+    setColor (color) {
+        this.color = color;
+        if (this.shape) {
+            if (this.filled) {
+                this.shape.fillColor = this.color;
+            } else {
+                this.shape.strokeColor = this.color;
+            }
+        }
+    }
+    setFilled (filled) {
+        if (this.filled === filled) return;
+        this.filled = filled;
+        if (this.shape && this.shape.isInserted()) {
+            if (this.filled) {
+                this.shape.fillColor = this.color;
+                this.shape.strokeWidth = 0;
+                this.shape.strokeColor = null;
+            } else {
+                this.shape.fillColor = null;
+                this.shape.strokeWidth = this.thickness;
+                this.shape.strokeColor = this.color;
+            }
+            this.onUpdateImage();
+        }
+    }
+    setThickness (thickness) {
+        if (this.thickness === thickness * paper.view.zoom) return;
+        this.thickness = thickness * paper.view.zoom;
+        if (this.shape && this.shape.isInserted() && !this.filled) {
+            this.shape.strokeWidth = this.thickness;
+        }
+        if (this.shape && this.shape.isInserted()) {
+            this.shape.data.zoomLevel = paper.view.zoom;
+            this.onUpdateImage();
+        }
+    }
+    handleMouseDown (event) {
+        if (event.event.button > 0) return; // only first mouse button
+        this.active = true;
+
+        if (this.boundingBoxTool.onMouseDown(
+            event, false /* clone */, false /* multiselect */, false /* doubleClicked */, this.getHitOptions())) {
+            this.isBoundingBoxMode = true;
+        } else {
+            this.isBoundingBoxMode = false;
+            clearSelection(this.clearSelectedItems);
+            this.commitShape();
+
+            const shapeOptions = {
+                point: event.downPoint,
+                strokeScaling: false,
+                size: 0,
+                strokeJoin: 'round'
```

Comment:
Now that I think about it, rectangles having rounded joins might be a bug. They were square in 2.0 and they're square in the vector editor. That would probably want to be in its own PR anyway if we wanted to fix that.